### DISCUSSION TITLE
Interactive resize in between windows (changing the ratio of two windows)

### DISCUSSION THREAD

**Discussion by elkowar** [YaLTeR/niri#868] [Ideas]
It'd be really nice to have interactive resizing have an option to, instead of just changing the width of the window you're targetting, also change the width of the window next to it, the same way that resizing windows vertically works inside a column.

Imagine you have two windows open, tiled to 1/3 and 2/3rds of the screen respectively. Now you want to see a bit more the left window, so you grab the right side of the window and drag it rightwards. Now the right window is moved outside of the viewport, which is rarely desired.

My proposal would be to have a hitbox _in between_ windows that makes it such that interactive affects both windows equally, changing the ratio of them.

This hitbox should optimally have a configurable width that takes precedence over the current resize behavior hitbox for dragging from the side of a window -- that'd allow people to have this "grab-inbetween-windows" resize even if they have no gaps in the window.
![image](https://github.com/user-attachments/assets/32fa5316-d003-4b6d-b023-fecd098caa84)

**Comment by YaLTeR** [comment]
Interesting idea, but the trigger mechanism needs to be something different I'm afraid. A lot of windows trigger CSD resize with cursor slightly outside the window geometry, but some trigger CSD resize only when the cursor is inside, in extreme cases there's only 1 px for the CSD resize. I don't think you can make the middle hitbox idea work reliably with all this variety while letting you not use it if you don't want to. 

Maybe it should be resizing with a modifier. Maybe it should be starting resize with a modifier held. 

There are other questions too, like what if the columns have multiple windows and they have different min/max width constraints?

**Reply by elkowar** [reply]
Isn't the question of differing min-max constraints for windows in columns an issue regardless of this feature?

And sure, having this be a different modifier also works well, I don't think I'd personally ever want the current behavior if this alternative was available 🤷🏻‍♂️ 

Maybe even just have a toggleable option for this version of resize vs. the current one.

I think this feature should just use the same system as the current modifier+rightclick resize behavior -- which should work regardless of CSD resize, no? So there, shouldn't the trigger of being in between windows be fine?

**Reply by YaLTeR** [reply]
> Isn't the question of differing min-max constraints for windows in columns an issue regardless of this feature?

Kind of, but here it's more complex since two columns are involved at once and you want to stop the resize once either of the column hits its width constraint.

> I don't think I'd personally ever want the current behavior if this alternative was available 🤷🏻‍♂️

If you resize a maximized window you probably don't want it synchronized with a random window that happened to be open to the right of it? What if it's a 98% wide window instead of maximized, where do you draw the line?

> I think this feature should just use the same system as the current modifier+rightclick resize behavior -- which should work regardless of CSD resize, no?

Mod+RMB does not use CSD, correct. But if it's this useful, would be a shame to limit it to Mod+RMB resize.

**Reply by phisch** [reply]
> Mod+RMB does not use CSD, correct. But if it's this useful, would be a shame to limit it to Mod+RMB resize.

I still hope that we get better ways to bind things. I'd like to define myself how to resize windows, maybe with `R+RMB`, and also binding to key-up events.

**Reply by elkowar** [reply]
> If you resize a maximized window you probably don't want it synchronized with a random window that happened to be open to the right of it? What if it's a 98% wide window instead of maximized, where do you draw the line?

I think I actually would, as in, that'd be the intuitive behavior -- and I don't see many cases where you'd want to increase the size of a fullscreen window either way, no? If enabled, this setting should just mean that resizing a window at a side that has a neighbouring accordingly resize the other widow as well.

> Kind of, but here it's more complex since two columns are involved at once and you want to stop the resize once either of the column hits its width constraint.

Fair, but that seems more like implementation complexity than design complexity, or am I missing something there?

> Mod+RMB does not use CSD, correct. But if it's this useful, would be a shame to limit it to Mod+RMB resize.

very valid, although I'd already be very happy with having it be only available on mod+RMB resize personally -- and _also_ having support for a modifier being held when doing CSD resize seems like a nice additional feature.

**Reply by YaLTeR** [reply]
Well, I guess need to try implementing it and see what problems come up and how it feels
